engine/ui/ui.go

```diff
@@ -0,0 +1,332 @@
+			opts = append(opts, useExistingBuildFromSource)
+		}
+		opts = append(opts, answerLatestRelease)
+	} else {
+		opts = append(opts, answerVersionRelease)
+	}
+
+	opts = append(opts, answerDoNothing)
+
+	ask := fmt.Sprintf("What do you want to do?")
+
+	selected := cli.MultiChoice(ask, opts...)
+
+	switch opts[selected] {
+	case answerDoNothing:
+		return nil
+	case answerLatestRelease:
+		return s.downloadStaticFilesFromGitHub("latest")
+	case answerVersionRelease:
+		return s.downloadStaticFilesFromGitHub(version)
+	case answerBuildFromSource:
+		return s.buildFromSource()
+	case useExistingBuildFromSource:
+		s.HTMLDir = "../ui/dist"
+	}
+	return nil
+}
+
+func (s *Service) buildFromSource() error {
+	if _, err := os.Stat("../ui"); os.IsNotExist(err) {
+		return fmt.Errorf("ui> You must have the directory ../ui with the ui source code")
+	}
+
+	if err := s.execCommand("npm install --no-audit"); err != nil {
+		return err
+	}
+	if err := s.execCommand("node --max_old_space_size=6000 node_modules/@angular/cli/bin/ng build --prod"); err != nil {
+		return err
+	}
+	s.HTMLDir = "../ui/dist"
+	return nil
+}
+
+func (s *Service) execCommand(command string) error {
+	log.Info("ui> running %s...", command)
+	parts := strings.Fields(command)
+	cmd := exec.Command(parts[0], parts[1:]...)
+	cmd.Dir = "../ui"
+
+	stderr, _ := cmd.StderrPipe()
+	cmd.Start()
+
+	scanner := bufio.NewScanner(stderr)
+	scanner.Split(bufio.ScanWords)
+	for scanner.Scan() {
+		m := scanner.Text()
+		fmt.Print(m + " ")
+	}
+	fmt.Println()
+	cmd.Wait()
+	return nil
+}
+
+func (s *Service) downloadStaticFilesFromGitHub(version string) error {
+	if _, err := os.Stat(s.Cfg.Staticdir); os.IsNotExist(err) {
+		log.Info("ui> creating directory %s", s.Cfg.Staticdir)
+		if err := os.Mkdir(s.Cfg.Staticdir, 0740); err != nil {
+			return fmt.Errorf("Error while creating directory: %v", err)
+		}
+	}
+
+	urlFiles := fmt.Sprintf("https://github.com/ovh/cds/releases/download/%s/ui.tar.gz", version)
+	if version == "latest" {
+		var err error
+		urlFiles, err = s.Client.DownloadURLFromGithub("ui.tar.gz")
+		if err != nil {
+			return fmt.Errorf("ui> Error while getting ui.tar.gz from Github err:%s", err)
+		}
+	}
+
+	log.Info("ui> Downloading from %s...", urlFiles)
+
+	resp, err := http.Get(urlFiles)
+	if err != nil {
+		return fmt.Errorf("Error while getting ui.tar.gz from GitHub: %v", err)
+	}
+	defer resp.Body.Close()
+
+	if err := sdk.CheckContentTypeBinary(resp); err != nil {
+		return fmt.Errorf(err.Error())
+	}
+
+	if resp.StatusCode != 200 {
+		return fmt.Errorf("Error while getting ui.tar.gz from GitHub - HTTP code: %d", resp.StatusCode)
+	}
+
+	log.Info("ui> Download in success, we are decompressing the archive now...")
+
+	return untarGZ(s.Cfg.Staticdir, resp.Body)
+}
+
+// untarGZ takes a destination path and a reader; a tar reader loops over the tarfile
+// creating the file structure at 'dst' along the way, and writing any files
+func untarGZ(dst string, r io.Reader) error {
```

Comment:
note sure, but i think we have a lib for that

Reply:
yep